File: DateDrop3/ViewModels/QuestionnaireViewModel.swift
//
//  QuestionnaireViewModel.swift
//  DateDrop3
//
//  问卷视图模型
//

import Foundation
import Combine

class QuestionnaireViewModel: ObservableObject {
    @Published var questions: [Question] = []
    @Published var sections: [String: [Question]] = [:]
    @Published var answers: [Int: Int] = [:]  // question_id: answer
    @Published var isLoading = false
    @Published var isSubmitting = false
    @Published var loadErrorMessage: String?
    @Published var submitErrorMessage: String?
    @Published var submitSuccess = false
    @Published var answerStatus: AnswerStatusResponse?

    private let sectionOrder = ["core_values", "lifestyle", "political", "relationship", "personality", "communication"]
    private let sectionNames = [
        "core_values": "核心价值观",
        "lifestyle": "生活方式",
        "political": "政治观点",
        "relationship": "关系期望",
        "personality": "性格特质",
        "communication": "沟通模式"
    ]
    private let sectionNamesEn = [
        "core_values": "Core Values",
        "lifestyle": "Lifestyle",
        "political": "Politics",
        "relationship": "Relationship",
        "personality": "Personality",
        "communication": "Communication"
    ]

    private var cancellables = Set<AnyCancellable>()
    private let networkManager = NetworkManager.shared
    private let questionsCacheKeyPrefix = "cached_questions_sections_"
    private let questionsCacheTTL: TimeInterval = 60 * 60 * 24
    private let answerStatusCacheKey = "cached_answer_status"
    private let answerStatusCacheTTL: TimeInterval = 60 * 5

    private struct QuestionsCache: Codable {
        let questions: [Question]
        let sections: [String: [Question]]
        let updatedAt: TimeInterval
    }

    private struct AnswerStatusCache: Codable {
        let status: AnswerStatusResponse
        let updatedAt: TimeInterval
    }

    static func clearAnswerStatusCache() {
        UserDefaults.standard.removeObject(forKey: "cached_answer_status")
    }

    private var appLanguage: String {
        UserDefaults.standard.string(forKey: "app_language") ?? "en"
    }

    // MARK: - 加载问卷

    func loadQuestions(lang: String = "en") {
        loadErrorMessage = nil

        if let cache = loadQuestionsCache(lang: lang) {
            questions = cache.questions
            sections = cache.sections
            isLoading = false
            if isQuestionsCacheFresh(cache) {
                return
            }
        } else {
            isLoading = true
        }
        
        networkManager.getQuestionsBySection(lang: lang)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case let .failure(error) = completion {
                        self?.loadQuestionsFallback(lang: lang, originalError: error)
                    }
                },
                receiveValue: { [weak self] response in
                    guard let self = self else { return }
                    self.isLoading = false
                    // 动态合并所有 section 下的题目
                    let combined = response.sections.values.flatMap { $0 }
                    self.questions = combined
                    self.sections = response.sections
                    self.saveQuestionsCache(lang: lang, questions: combined, sections: response.sections)
                }
            )
            .store(in: &cancellables)
    }
    
    private func loadQuestionsFallback(lang: String, originalError: Error) {
        networkManager.getQuestions(lang: lang)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    guard let self = self else { return }
                    self.isLoading = false
                    if case let .failure(error) = completion {
                        if self.appLanguage == "zh" {
                            self.loadErrorMessage = "加载题目失败: \(originalError.localizedDescription); 备用接口错误: \(error.localizedDescription)"
                        } else {
                            self.loadErrorMessage = "Failed to load questions: \(originalError.localizedDescription); fallback error: \(error.localizedDescription)"
                        }
                    }
                },
                receiveValue: { [weak self] response in
                    guard let self = self else { return }
                    self.loadErrorMessage = nil
                    self.questions = response.questions
                    // 尝试按 section 分组, 没有则归类到 unknown
                    let grouped = Dictionary(grouping: response.questions, by: { $0.section ?? "unknown" })
                    self.sections = grouped
                    self.saveQuestionsCache(lang: lang, questions: response.questions, sections: grouped)
                }
            )
            .store(in: &cancellables)
    }

    // MARK: - 答案管理

    func setAnswer(questionId: Int, answer: Int) {
        answers[questionId] = answer
    }

    func getAnswer(questionId: Int) -> Int? {
        return answers[questionId]
    }

    func getProgress() -> (answered: Int, total: Int) {
        let answered = answers.count
        return (answered, 66)
    }

    func isQuestionAnswered(questionId: Int) -> Bool {
        return answers[questionId] != nil
    }

    func getSectionProgress(section: String) -> (answered: Int, total: Int) {
        guard let sectionQuestions = sections[section] else {
            return (0, 0)
        }

        let answered = sectionQuestions.filter { isQuestionAnswered(questionId: $0.id) }.count
        return (answered, sectionQuestions.count)
    }

    // MARK: - 提交答案

    func submitAnswers() {
        isSubmitting = true
        submitErrorMessage = nil
        submitSuccess = false

        // 验证是否全部答完
        if answers.count != 66 {
            submitErrorMessage = appLanguage == "zh" ? "请完成全部66道题目" : "Please complete all 66 questions"
            isSubmitting = false
            return
        }

        // 构建答案数组
        let answerArray: [Answer] = answers.map { questionId, answer in
            Answer(question_id: questionId, answer: answer)
        }.sorted { $0.question_id < $1.question_id }

        networkManager.submitAnswers(answers: answerArray)
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    self?.isSubmitting = false
                    if case let .failure(error) = completion {
                        self?.submitErrorMessage = error.localizedDescription
                    }
                },
                receiveValue: { [weak self] response in
                    self?.submitSuccess = true
                    self?.submitErrorMessage = nil
                    self?.loadAnswerStatus(force: true)
                }
            )
            .store(in: &cancellables)
    }

    // MARK: - 答案状态

    func loadAnswerStatus(force: Bool = false) {
        if !force, let cache = loadAnswerStatusCache(), isAnswerStatusCacheFresh(cache) {
            answerStatus = cache.status
            return
        }
        networkManager.getAnswerStatus()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { _ in },
                receiveValue: { [weak self] status in
                    self?.answerStatus = status
                    self?.saveAnswerStatusCache(status)
                }
            )
            .store(in: &cancellables)
    }

    // MARK: - 辅助方法

    func getSectionName(_ section: String) -> String {
        if appLanguage == "zh" {
            return sectionNames[section] ?? section
        }
        return sectionNamesEn[section] ?? section
    }

    func getSectionOrder() -> [String] {
        return sectionOrder
    }

    func getQuestionsForSection(_ section: String) -> [Question] {
        return sections[section] ?? []
    }

    private func saveQuestionsCache(lang: String, questions: [Question], sections: [String: [Question]]) {
        let cache = QuestionsCache(questions: questions, sections: sections, updatedAt: Date().timeIntervalSince1970)
        if let encoded = try? JSONEncoder().encode(cache) {
            UserDefaults.standard.set(encoded, forKey: questionsCacheKeyPrefix + lang)
        }
    }

    private func loadQuestionsCache(lang: String) -> QuestionsCache? {
        guard let data = UserDefaults.standard.data(forKey: questionsCacheKeyPrefix + lang) else {
            return nil
        }
        return try? JSONDecoder().decode(QuestionsCache.self, from: data)
    }

    private func isQuestionsCacheFresh(_ cache: QuestionsCache) -> Bool {
        Date().timeIntervalSince1970 - cache.updatedAt < questionsCacheTTL
    }

    private func saveAnswerStatusCache(_ status: AnswerStatusResponse) {
        let cache = AnswerStatusCache(status: status, updatedAt: Date().timeIntervalSince1970)
        if let encoded = try? JSONEncoder().encode(cache) {
            UserDefaults.standard.set(encoded, forKey: answerStatusCacheKey)
        }
    }

    private func loadAnswerStatusCache() -> AnswerStatusCache? {
        guard let data = UserDefaults.standard.data(forKey: answerStatusCacheKey) else {
            return nil
        }
        return try? JSONDecoder().decode(AnswerStatusCache.self, from: data)
    }

    private func isAnswerStatusCacheFresh(_ cache: AnswerStatusCache) -> Bool {
        Date().timeIntervalSince1970 - cache.updatedAt < answerStatusCacheTTL
    }
}

File: app-android/app/src/main/java/com/vitaduo/datedrop/viewmodel/QuestionnaireViewModel.kt
package com.vitaduo.datedrop.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vitaduo.datedrop.model.Answer
import com.vitaduo.datedrop.model.AnswerSubmitRequest
import com.vitaduo.datedrop.model.Question
import com.vitaduo.datedrop.model.AnswerStatusResponse
import com.vitaduo.datedrop.network.NetworkManager
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Locale

class QuestionnaireViewModel : ViewModel() {
    private val _questions = MutableStateFlow<List<Question>>(emptyList())
    val questions: StateFlow<List<Question>> = _questions.asStateFlow()

    private val _sections = MutableStateFlow<Map<String, List<Question>>>(emptyMap())
    val sections: StateFlow<Map<String, List<Question>>> = _sections.asStateFlow()

    private val _answers = MutableStateFlow<Map<Int, Int>>(emptyMap())
    val answers: StateFlow<Map<Int, Int>> = _answers.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _isSubmitting = MutableStateFlow(false)
    val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()

    private val _loadErrorMessage = MutableStateFlow<String?>(null)
    val loadErrorMessage: StateFlow<String?> = _loadErrorMessage.asStateFlow()

    private val _submitErrorMessage = MutableStateFlow<String?>(null)
    val submitErrorMessage: StateFlow<String?> = _submitErrorMessage.asStateFlow()

    private val _submitSuccess = MutableStateFlow(false)
    val submitSuccess: StateFlow<Boolean> = _submitSuccess.asStateFlow()

    private val sectionOrder = listOf("core_values", "lifestyle", "political", "relationship", "personality", "communication")
    private val sectionNames = mapOf(
        "core_values" to "核心价值观",
        "lifestyle" to "生活方式",
        "political" to "政治观点",
        "relationship" to "关系期望",
        "personality" to "性格特质",
        "communication" to "沟通模式"
    )
    private val sectionNamesEn = mapOf(
        "core_values" to "Core Values",
        "lifestyle" to "Lifestyle",
        "political" to "Politics",
        "relationship" to "Relationship",
        "personality" to "Personality",
        "communication" to "Communication"
    )

    fun loadQuestions() {
        val lang = NetworkManager.getAppLanguage()
        viewModelScope.launch {
            _isLoading.value = true
            _loadErrorMessage.value = null
            try {
                val response = NetworkManager.api.getQuestionsBySection(lang)
                val sectionMap = response.sections ?: response.data?.sections ?: emptyMap()
                _sections.value = sectionMap
                _questions.value = sectionMap.values.flatten()
            } catch (e: Exception) {
                loadQuestionsFallback(lang, e.message)
            } finally {
                _isLoading.value = false
            }
        }
    }

    private suspend fun loadQuestionsFallback(lang: String, originalError: String?) {
        try {
            val response = NetworkManager.api.getQuestions(lang)
            _questions.value = response.questions
            _sections.value = response.questions.groupBy { it.section ?: "unknown" }
        } catch (e: Exception) {
            _loadErrorMessage.value = "Failed to load questions: $originalError; Fallback error: ${e.message}"
        }
    }

    fun setAnswer(questionId: Int, answer: Int) {
        val currentAnswers = _answers.value.toMutableMap()
        currentAnswers[questionId] = answer
        _answers.value = currentAnswers
    }

    fun submitAnswers() {
        val appLanguage = NetworkManager.getAppLanguage()
        if (_answers.value.size != 66) {
            _submitErrorMessage.value = if (appLanguage == "zh") "请完成全部66道题目" else "Please complete all 66 questions"
            return
        }

        viewModelScope.launch {
            _isSubmitting.value = true
            _submitErrorMessage.value = null
            _submitSuccess.value = false
            try {
                val answerList = _answers.value.map { Answer(it.key, it.value) }.sortedBy { it.questionId }
                NetworkManager.api.submitAnswers(AnswerSubmitRequest(answerList))
                _submitSuccess.value = true
            } catch (e: Exception) {
                _submitErrorMessage.value = e.message
            } finally {
                _isSubmitting.value = false
            }
        }
    }

    fun getSectionName(section: String): String {
        val appLanguage = NetworkManager.getAppLanguage()
        return if (appLanguage == "zh") sectionNames[section] ?: section else sectionNamesEn[section] ?: section
    }

    fun clearLoadError() {
        _loadErrorMessage.value = null
    }

    fun clearSubmitError() {
        _submitErrorMessage.value = null
    }

    fun getSectionOrder() = sectionOrder

    fun getProgress(): Pair<Int, Int> {
        return _answers.value.size to 66
    }
}
